### ISSUE TITLE
InputScope does not recognize .Del key on the numberpad.

### ISSUE THREAD

**Issue by tlckfeks** [suurjaak/InputScope#14]
Hello, Developer~!
On my keyboard, InputScope does not recognize the .Del key on the numberpad.
What settings do I need to modify?
Thank you.

**Comment by suurjaak** [comment]
Does it recognize it as anything at all? E.g. if you press Del many times, do statistics change on the keyboard page?

**Comment by tlckfeks** [comment]
> Does it recognize it as anything at all? E.g. if you press Del many times, do statistics change on the keyboard page?

I pressed the .Del key dozens of times, but no logs were saved in the InputScope. :(
InputScope doesn't recognize .Del key at all.

**Comment by suurjaak** [comment]
A few more questions:

1. What is the operating system?
2. How was InputScope installed?
3. What is the keyboard model?
4. What is the keyboard layout?

**Comment by tlckfeks** [comment]
1. Windows 7 x64
2. InputScope_1.5_x64.exe in C:\Program Files\InputScope
3. Membrane PS/2 Samsung Keyboard. 
4. http://img.danawa.com/prod_img/large/group_249/124741_1.jpg      
http://img.danawa.com/prod_img/large/group_68/34147_1.jpg

**Comment by suurjaak** [comment]
To clarify: by keyboard layout I meant the OS keyboard language setting.

**Comment by tlckfeks** [comment]
KOREAN - Microsoft IME
PS/2 Keyboard 103/106KEY

**Comment by suurjaak** [comment]
If you switch the layout to something else like English, does the result change?

**Comment by tlckfeks** [comment]
I switched to "English (United States) - US keyboard" but the result is unchanged.

**Comment by tlckfeks** [comment]
FYI, The Mousotron recognizes .Del key normally.
https://www.blacksunsoftware.com/mousotron.html

**Comment by suurjaak** [comment]
Can you try the following.

- right-click on the taskbar icon and select "Show Python console"
- in the console, copy-paste the following lines one by one, and press Enter after each:
```python
import pynput

listener = pynput.keyboard.Listener(on_press=lambda k: print("%s -- %s -- %s" % (type(k), vars(k), k)))

listener.start()
```

If you press the Del key, what does it print?

(The printing can be stopped with running `listener.stop()` in the console.)

**Comment by tlckfeks** [comment]
<class 'pynput.keyboard._win32.KeyCode'> -- {'vk': 110, 'char': None, 'is_dead': False, 'combining': None, '_flags': None, '_scan': 83} -- <110>

**Comment by tlckfeks** [comment]
um... I found that Inputscope does not recognize the combo of ctrl and number keys. For example, ctrl+1 or ctrl+alt+0.

**Comment by suurjaak** [comment]
Confirmed. Will release a fixed version soon.

**Comment by tlckfeks** [comment]
It’s such a relief to know that you fixed this issue.

um... There are 2 more special keys that inputscope does not recognize.

All Koreans use two special keys.

https://images2.imgbox.com/51/f5/nrKzSAKT_o.jpg 

<class 'pynput.keyboard._win32.KeyCode'> -- {'vk': 25, 'char': None, 'is_dead': False, 'combining': None, '_flags': None, '_scan': 241} -- <25>
Key Name : ' Hanja '
Function : Replace Korean with Classical Chinese

<class 'pynput.keyboard._win32.KeyCode'> -- {'vk': 21, 'char': None, 'is_dead': False, 'combining': None, '_flags': None, '_scan': 242} -- <21>
Key name : ' KO/EN '
Function : Language switching (Korean ↔ English) 

If you don’t mind, could you add those to inputscope too?

**Comment by suurjaak** [comment]
Regarding Korean keys, sure, I can try.

The second key, is that the one also known as [Han/Yeong](https://en.wikipedia.org/wiki/Language_input_keys#Han/Yeong_(%ED%95%9C/%EC%98%81)) ?

**Comment by tlckfeks** [comment]
It is true that the Korean name for the second key is only called 'Han/Yeong' by Koreans.
But In my opinion, I think it's okay to call it Korean/English.
Whatever you want to call it is fine. :)

**Comment by suurjaak** [comment]
Can you try this in the console:

```python
import pynput

on_press = lambda k: print("%s -- %s -- %s" % (type(k), vars(k), k))

w32filter = lambda a, b: print("<RAW virt %r scan %r flag %r>" % (b.vkCode, b.scanCode, b.flags)) if a == 256 else None

listener = pynput.keyboard.Listener(on_press=on_press, win32_event_filter=w32filter)

listener.start()
```

What does it print for the Korean specials?

**Comment by tlckfeks** [comment]
Hanja : <RAW virt 25 scan 241 flag 1>
Han/Yeong : <RAW virt 21 scan 242 flag 1>

**Comment by suurjaak** [comment]
Can you try downloading and running https://erki.lap.ee/stuff/inputscope_1.6.dev8_x64.exe and see how it fares? 

Numpad-Delete and the Korean keys should register now, and Ctrl+number combos as well.

I named the Korean keys as "IME Hanja/Kanji" and "IME Hangul/Kana".

**Comment by tlckfeks** [comment]
Numpad-Delete	1
Ctrl-9                       1
IME Kana/Hangul	2
IME Hanja/Kanji	3

It works like a charm!
I thought these were simple issues, but it seems to have been a longer journey than I thought.
Thank you for your hard work :)

By the way, Kana/Hangul?
As far as I know, Kana means Katakana in Japanese.
Korean keyboards do not have the special key to switch between Hangul and Katakana.
Han/Yeong only supports switching between Hangul and English.

And the Virustotal seems to recognize the dev version as a virus.
https://www.virustotal.com/gui/file/8b4f7ca70642e206cba75761d499b08e912fdc016fb68f42712c419a9e349413?nocache=1

**Comment by suurjaak** [comment]
Excellent that it all works now. I will make a public release soon; will keep this issue open until then. These were simple problems indeed, but I also had some other changes in the pipeline pending release. 

I chose more generic names for the language keys, as I discovered that the same keys are used for similar language switching on Japanese keyboards as well (https://docs.microsoft.com/en-us/windows/win32/inputdev/virtual-key-codes). The name does not mean it switches between Kana and Hangul, but that it can toggle Hangul mode if Korean, or it can toggle Kana mode if Japanese. 

By the way, I added an option for users to specify their own additional keys or override key names in the configuration file. So for example you can add to `inputscope.ini`:

```javascript
CustomKeys = {21: "IME Han/Yeong", 25: "IME Hanja"}
```

---

The antivirus alerts, are, unfortunately, an eternal and expected pain. There is something about packaged Python apps that *always* triggers some antivirals, possibly because the methods for packing the executable code are similar to what some viruses use. And it is quite hopeless to get it all green; the false positives need to be reported to each antivirus separately, and some of them do not even have channels for this. The current report state is actually better than usual, only some rather no-name antivirals are raising an alert. 

Ironic that Yandex and Kaspersky are still used in VirusTotal. The entire Russian Federation has been sanctioned by the western world in all possible ways, but hey, let's still use their antivirals..

**Comment by tlckfeks** [comment]
Alright, I understood the whole process.
I'm looking forward to the release of the next stable version. :)
Have a nice day!

**Comment by suurjaak** [comment]
Released v1.6, the problems above ought to be fixed now.
